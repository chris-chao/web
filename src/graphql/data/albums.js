// @flow
import { arrayPropertyDescriptor, makeMap, propertyDescriptor } from './utils';
import { songs, artists } from '.';
import type { Song, Artist } from '.';

type Source = {
  id: string,
  name: string,
  artworkUrl: string,
  artistId: string,
  songIds: string[],
  releaseYear: number,
};

const albums: Source[] = [
  {
    id: '0d9396f5-5454-47ce-9bd8-dacaf2fbf6d0',
    name: 'Stole the Show',
    artworkUrl:
      'https://i.scdn.co/image/d345ab2a8278434f1c8cc936ace70da02ac845fb',
    artistId: 'b4976c45-3362-4e50-9274-63a375b3a299',
    songIds: ['69120ac9-1e48-494f-a1f4-4a34735fe408'],
    releaseYear: 2015,
  },
  {
    id: '44db420c-0995-4a74-9cc2-0191d65bfdd0',
    name: "I'm That (Remix)",
    artistId: 'c1e012c3-f496-41e9-aa67-a060b666bd78',
    songIds: ['9283bed7-a183-4f54-a0d7-820946c220e3'],
    releaseYear: 2015,
    artworkUrl:
      'http://is4.mzstatic.com/image/thumb/Music5/v4/08/da/96/08da9619-3f9b-7c95-60d1-6c18cfdd4dbd/source/600x600bb.jpg',
  },
  {
    id: 'd0941f59-2138-433c-9ccf-368040228593',
    name: 'Bugatti',
    artworkUrl:
      'https://upload.wikimedia.org/wikipedia/en/3/3f/AceHood_Bugatti.jpg',
    artistId: '79506046-e727-405b-924a-2daf428e772f',
    songIds: ['492667ee-5bed-45ff-8a2c-67277677cf86'],
    releaseYear: 2013,
  },
  {
    id: 'fc1acc4a-8e76-493c-ae3b-e71a88d41bbf',
    name: 'The Beautiful & Damned',
    artworkUrl:
      'https://upload.wikimedia.org/wikipedia/en/3/31/G-Eazy_-_The_Beautiful_%26_Damned.png',
    artistId: '04f6d289-154c-4602-9a33-e27d4be265d7',
    songIds: [],
    releaseYear: 2017,
  },
];

export type Album = {
  duration: number,
  artist: Artist,
  songs: Song[],
} & Source;

const processedAlbums: Map<string, Album> = makeMap(
  albums.map((album: Source): Album =>
    // Flow doesn't completely understand defineProperties.
    // https://github.com/facebook/flow/issues/285
    // $ExpectError
    Object.defineProperties(
      {
        ...album,
      },
      {
        duration: {
          get() {
            return (this: Album).songs
              .map(({ duration = 0 } = {}) => duration)
              .reduce((a, b) => a + b, 0);
          },
        },

        artist: propertyDescriptor(() => artists, album.artistId),
        songs: arrayPropertyDescriptor(() => songs, album.songIds),
      }
    )
  )
);

export default processedAlbums;
