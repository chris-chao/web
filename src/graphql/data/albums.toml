[[albums]]
id = 'album:1'
name = 'Stole the Show'
artworkUrl = 'https://i.scdn.co/image/d345ab2a8278434f1c8cc936ace70da02ac845fb'
artistId = 'artist:1'
songIds = ['song:1:1']
releaseYear = 2015

[[albums]]
id = 'album:2'
name = "I'm That (Remix)"
artistId = 'artist:2'
songIds = ['song:2:1']
releaseYear = 2015
artworkUrl = 'http://is4.mzstatic.com/image/thumb/Music5/v4/08/da/96/08da9619-3f9b-7c95-60d1-6c18cfdd4dbd/source/600x600bb.jpg'

[[albums]]
id = 'album:3'
name = 'Bugatti'
artworkUrl = 'https://upload.wikimedia.org/wikipedia/en/3/3f/AceHood_Bugatti.jpg'
artistId = 'artist:3'
songIds = ['song:3:1']
releaseYear = 2013

[[albums]]
id = 'album:4'
name = 'The Beautiful & Damned'
artworkUrl = 'https://upload.wikimedia.org/wikipedia/en/3/31/G-Eazy_-_The_Beautiful_%26_Damned.png'
artistId = 'artist:4'
songIds = []
releaseYear = 2017
