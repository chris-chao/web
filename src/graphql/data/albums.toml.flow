// @flow
import type { AlbumSource } from '../mock/albums';

declare export var albums: AlbumSource[];
