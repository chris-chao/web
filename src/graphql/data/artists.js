// @flow
import { makeMap } from './utils';

export type Artist = {
  id: string,
  name: string,
  albums: string[],
};

const artists: Artist[] = [
  {
    id: 'b4976c45-3362-4e50-9274-63a375b3a299',
    name: 'Kygo',
    albums: ['0d9396f5-5454-47ce-9bd8-dacaf2fbf6d0'],
  },
  {
    id: 'c1e012c3-f496-41e9-aa67-a060b666bd78',
    name: 'R. City',
    albums: ['44db420c-0995-4a74-9cc2-0191d65bfdd0'],
  },
  {
    id: '79506046-e727-405b-924a-2daf428e772f',
    name: 'Ace Hood',
    albums: ['d0941f59-2138-433c-9ccf-368040228593'],
  },
  {
    id: '04f6d289-154c-4602-9a33-e27d4be265d7',
    name: 'G-Eazy',
    albums: ['fc1acc4a-8e76-493c-ae3b-e71a88d41bbf'],
  },
];

const processedArtists: Map<string, Artist> = makeMap(artists);

export default processedArtists;
