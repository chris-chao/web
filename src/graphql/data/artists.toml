[[artists]]
id = 'artist:1'
name = 'Kygo'
albumIds = ['album:1']

[[artists]]
id = 'artist:2'
name = 'R. City'
albumIds = ['album:2']

[[artists]]
id = 'artist:3'
name = 'Ace Hood'
albumIds = ['album:3']

[[artists]]
id = 'artist:4'
name = 'G-Eazy'
albumIds = ['album:4']

[[artists]]
id = 'artist:5'
name = 'Zoe Nash'
albumIds = []

[[artists]]
id = 'artist:6'
name = 'Halsey'
albumIds = []

[[artists]]
id = 'artist:7'
name = 'Charlie Puth'
albumIds = []

[[artists]]
id = 'artist:8'
name = 'ASAP Rocky'
albumIds = []

[[artists]]
id = 'artist:9'
name = 'Cardi B'
albumIds = []

[[artists]]
id = 'artist:10'
name = 'Anna of the North'
albumIds = []

[[artists]]
id = 'artist:11'
name = 'Leviathan'
albumIds = []

[[artists]]
id = 'artist:12'
name = 'Kehlani'
albumIds = []

[[artists]]
id = 'artist:13'
name = 'E-40'
albumIds = []

[[artists]]
id = 'artist:14'
name = 'Jay Ant'
albumIds = []

[[artists]]
id = 'artist:15'
name = 'SG Lewis'
albumIds = []

[[artists]]
id = 'artist:16'
name = 'Louis Mattrs'
albumIds = []

[[artists]]
id = 'artist:17'
name = 'Madison Love'
albumIds = []

[[artists]]
id = 'artist:18'
name = 'Ugochi'
albumIds = []

[[artists]]
id = 'artist:19'
name = 'Drew Love'
albumIds = []

[[artists]]
id = 'artist:20'
name = 'Son Lux'
albumIds = []

[[artists]]
id = 'artist:21'
name = 'Beenie Man'
albumIds = []

[[artists]]
id = 'artist:22'
name = 'Azealia Banks'
albumIds = []

[[artists]]
id = 'artist:23'
name = 'Rick Ross'
albumIds = []

[[artists]]
id = 'artist:24'
name = 'Future'
albumIds = []
