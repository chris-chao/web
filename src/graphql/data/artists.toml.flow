// @flow
import type { ArtistSource } from '../mock/artists';

declare export var artists: ArtistSource[];
