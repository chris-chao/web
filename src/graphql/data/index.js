// @flow
export { default as albums } from './albums';
export { default as artists } from './artists';
export { default as songs } from './songs';
export { default as stats } from './stats';
export { default as playlists } from './playlists';

export type { Album } from './albums';
export type { Artist } from './artists';
export type { Song } from './songs';
export type { SongUserStats } from './stats';
export type { Playlist, PlaylistItem } from './playlists';
