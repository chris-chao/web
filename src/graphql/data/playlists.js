// @flow
import { makeMap, mustGetKeys } from './utils';
import { songs as songMap } from '.';
import type { Song } from '.';

type Source = {
  id: string,
  name: string,
  songIds: string[],
};

const playlists: Source[] = [
  {
    id: 'playlist:1',
    name: 'Quick Jam Session',
    songIds: [
      '9283bed7-a183-4f54-a0d7-820946c220e3',
      '492667ee-5bed-45ff-8a2c-67277677cf86',
    ],
  },
  {
    id: 'playlist:2',
    name: 'Chill and Netflix',
    songIds: ['69120ac9-1e48-494f-a1f4-4a34735fe408'],
  },
];

export type PlaylistItem = {
  id: string,
  song: Song,
};

export type Playlist = {
  items: PlaylistItem[],
  duration: number,
} & Source;

const map: Map<string, Playlist> = makeMap(
  playlists.map(playlist =>
    // $ExpectError
    Object.defineProperties(
      { ...playlist },
      {
        items: {
          get() {
            const songs: Song[] = mustGetKeys(songMap, playlist.songIds);
            return songs.map((song, index) => ({
              id: `${playlist.id}:${index}`,
              song: song,
            }));
          },
        },
        duration: {
          get() {
            return this.items
              .map(item => item.song.duration)
              .reduce((a, b) => a + b, 0);
          },
        },
      }
    )
  )
);

export default map;
