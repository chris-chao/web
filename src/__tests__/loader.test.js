// @flow
import data from './test.toml';

it('should load toml as an object', () => {
  expect(data).toMatchSnapshot();
});
