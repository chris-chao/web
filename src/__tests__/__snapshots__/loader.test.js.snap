// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should load toml as an object 1`] = `
Object {
  "thing": Array [
    Object {
      "key": "value",
    },
    Object {
      "key": "value 2",
    },
  ],
}
`;
